lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java

```diff
@@ -337,9 +354,79 @@ public MavenPluginManager getMavenPluginManager() {
 		WorkspaceReader workspaceReader = mavenRequest.getWorkspaceReader();
 		if (workspaceReader instanceof MavenLemminxWorkspaceReader) {
-			// TODO
+			Collection<DOMDocument> projectsToAdd = computeAddedWorkspaceProjects(added != null? added : new URI[0]);
+			Collection<URI> projectsToRemove = computeRemovedWorkspaceProjects(removed != null ? removed : new URI[0]);
+
+			if (projectsToAdd != null) {
+				projectsToAdd.stream().forEach(d -> getProjectCache().addDocument(URI.create(d.getDocumentURI()), d));
+			}
+
+			if (projectsToRemove != null) {
+				projectsToRemove.stream().forEach(u -> getProjectCache().removeDocument(u));
+			}
 		}
 	}
+	
+	private Collection<URI> computeRemovedWorkspaceProjects(URI[] removed) {
+		Collection<MavenProject> cachedProjects = getProjectCache().getProjects();
+		return cachedProjects.stream().filter(p -> {
+			return Arrays.asList(removed).stream().filter(uri -> {
+				Path removedPath = new File(uri).toPath();
+				return p.getFile().toPath().startsWith(removedPath);
+			}).findAny().isPresent();
+		}).map(p -> p.getFile().toURI()).collect(Collectors.toUnmodifiableList());
+	}
+	
+	private List<DOMDocument> computeAddedWorkspaceProjects(URI[] added) {
+		List<DOMDocument> projectsToAdd = new ArrayList<DOMDocument>();
+		Arrays.asList(added).stream().forEach(uri -> {
+			Path addedPath = new File(uri).toPath();
+			try {
+				Files.walkFileTree(addedPath, Collections.emptySet(), 10, new SimpleFileVisitor<Path>() {
+					@Override
+					public FileVisitResult preVisitDirectory(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().charAt(0) == '.') {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+						return FileVisitResult.CONTINUE;
+					}
+
+					@Override
+					public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().toLowerCase().endsWith(".xml")) {
+							DOMDocument document;
+							try {
+								document = createDOMDocument(file.toUri());
+								if (document != null) {
+									DOMElement documentElement = document.getDocumentElement();
+									if (documentElement != null && PROJECT_ELT.equals(documentElement.getLocalName())) {
+										projectsToAdd.add(document);
+									}
+								}
+							} catch (URISyntaxException e) {
+								LOGGER.log(Level.SEVERE, e.getMessage(), e);
+							}
+						}
+						return FileVisitResult.CONTINUE;
+					}
+				});
+			} catch (IOException e) {
+				LOGGER.log(Level.SEVERE, e.getMessage(), e);
+			}
+		});
+		
+		return projectsToAdd;
+	}
+	
+	private DOMDocument createDOMDocument(URI uri) throws IOException, URISyntaxException {
+		File file = new File(uri);
+		String contents = Files.readString(file.toPath());
+		return org.eclipse.lemminx.dom.DOMParser.getInstance()
+				.parse(new TextDocument(new TextDocumentItem(uri.toString(), "xml", 1, contents)), 
+						languageService.getResolverExtensionManager());

```

Comment:
As you don't need extensions at that stage, you can probably just use `new URIResolverExtensionManager()` so you don't need an XMLLanguageService.

Reply:
OK